src/server/p2p/network.rs

```diff
@@ -1429,61 +1429,65 @@ where S: ShareChain
-                    info!(target: SYNC_REQUEST_LOG_TARGET, "[{:?}] num catchups: {:?}", algo, num_catchups);
-                    if must_continue_sync && num_catchups.map(|n| n < MAX_CATCH_UP_ATTEMPTS).unwrap_or(true) {
-                        info!(target: SYNC_REQUEST_LOG_TARGET, "Continue catch up for {}", peer);
-                        peer_store_write_lock.add_catch_up_attempt(&peer);
-                        let perform_catch_up_sync = PerformCatchUpSync {
-                            algo,
-                            peer,
-                            last_block_from_them,
-                            their_height,
-                        };
-                        let _ = tx.send(InnerRequest::PerformCatchUpSync(perform_catch_up_sync));
-                    } else {
-                        info!(target: SYNC_REQUEST_LOG_TARGET, "Catch up sync completed for chain {} from {} after {} catchups", algo, peer, num_catchups.map(|s| s.to_string()).unwrap_or_else(|| "None".to_string()));
-
-                        // this only gets called after sync completes. So lets make sure we can mine since we completed
-                        // a sync
-                        synced_bool.store(true, std::sync::atomic::Ordering::Relaxed);
-                        peer_store_write_lock.reset_catch_up_attempts(&peer);
-                    }
-                },
-                Err(error) => match error {
-                    crate::sharechain::error::ShareChainError::BlockParentDoesNotExist { missing_parents } => {
-                        // This should not happen though, catchup should return all blocks
-                        warn!(target: SYNC_REQUEST_LOG_TARGET, squad; "Catchup sync Reporting missing blocks {}", missing_parents.len());
-                        let sync_share_chain = SyncShareChain {
-                            algo,
-                            peer,
-                            missing_parents,
-                            is_from_new_block_notify: false,
-                        };
-                        let _ = tx.send(InnerRequest::DoSyncChain(sync_share_chain));
-                        return;
+            for b in blocks {
+                match share_chain.add_synced_blocks(&[b]).await {
+                    Ok(result) => {
+                        info!(target: LOG_TARGET, "Block added");
                     },
-                    _ => {
-                        error!(target: SYNC_REQUEST_LOG_TARGET, squad; "Failed to add Catchup synced blocks to share chain: {error:?}");
-                        network_peer_store
-                            .write()
-                            .await
-                            .move_to_grey_list(peer, format!("Block failed validation: {error}"));
+                    Err(error) => match error {
+                        crate::sharechain::error::ShareChainError::BlockParentDoesNotExist { missing_parents } => {
```

Comment:
You should probably map this into a single request
Its possible that you might need a missing block that you already have